### PR TITLE
Allow managing Project Sources on Content Projects

### PR DESCRIPTION
**Pull request by hustodemon**
## What does this PR change?

^^^

## How to review?
Commit-by-commit.

## TODO
- [x] Revisit all the thrown exceptions and make them reasonable & consintent

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- XMLRPC api doc added

- [x] **DONE**

## Test coverage
- Unit tests were added

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/6538 and https://github.com/SUSE/spacewalk/issues/6513

- [x] **DONE**

### CODE REVIEW

**Comment by lneves12** [comment]
Just a reminder, don't forget about the new enum column on ProjectSource that we talked about to keep the sources changes

I would suggest something like:

0: "not edited"
1: "new"
2: "removed
3: "edited"

What do you think?

**Comment by hustodemon** [comment]
Rebased and added the explicit many-to-many project-filter entity.

File: java/code/src/com/redhat/rhn/domain/contentmgmt/ContentProject.java [hunk 1]
```diff
@@ -224,19 +224,30 @@ public void setSources(List<ProjectSource> sourcesIn) {
...
+     * @param source the Source
+     * @throws java.lang.IllegalArgumentException if the given Source does not match to the Project
+     * @return true if Source was added, false if the Source had been present already
      */
-    public void addSource(ProjectSource source) {
-        source.setContentProject(this);
-        sources.add(source);
+    public boolean addSource(ProjectSource source) {
+        if (!source.getContentProject().equals(this)) {
+            throw new IllegalArgumentException("Trying to attach source " + source + " with mismatched Project.");
```
Comment (ncounter) [inline java/code/src/com/redhat/rhn/domain/contentmgmt/ContentProject.java]: I am maybe too blind to figured this out but, what is the code-path/work-flow that lets this happen?

File: java/code/src/com/redhat/rhn/manager/contentmgmt/test/ContentManagerTest.java [hunk 2]
```diff
@@ -253,4 +273,113 @@ public void testEnvironmentPermissions() {
...
+        assertTrue(ContentManager.lookupProjectSource("cplabel", SW_CHANNEL, channel.getLabel(), user).isPresent());
+        ContentManager.removeProject("cplabel", user);
+        // we can't use ContentManager.lookupProjectSource because the project does not exist
+        assertTrue(HibernateFactory.getSession()
+                .createQuery("SELECT 1 FROM SoftwareProjectSource s where s.contentProject = :cp")
+                .setParameter("cp", cp)
+                .list()
+                .isEmpty());
+    }
+}
```
Comment (ncounter) [inline java/code/src/com/redhat/rhn/manager/contentmgmt/test/ContentManagerTest.java]: Blank new line missing

File: java/code/src/com/redhat/rhn/domain/contentmgmt/ProjectSource.java [hunk 3]
```diff
@@ -41,8 +45,107 @@
 public abstract class ProjectSource {
 
     private Long id;
+    private State state;
     private ContentProject contentProject;
 
+    /**
+     * State of the Source
```
Comment (ncounter) [inline java/code/src/com/redhat/rhn/domain/contentmgmt/ProjectSource.java]: Just to be nice with developers new to this feature, I would add a small description like:
```
ATTACHED, // softly added to the Project but not yet part of the Build output
DETACHED, // softly removed from the Project will be really removed after the next Project Build
BUILT, // part of the latest Build
```

File: java/code/src/com/redhat/rhn/domain/contentmgmt/ContentProject.java (hunk 1 above)
Comment (hustodemon) [reply java/code/src/com/redhat/rhn/domain/contentmgmt/ContentProject.java]: You're right. This is unnecessary. I added the for some reason when I implemented the positioning.

File: java/code/src/com/redhat/rhn/domain/contentmgmt/ProjectSource.java (hunk 3 above)
Comment (hustodemon) [reply java/code/src/com/redhat/rhn/domain/contentmgmt/ProjectSource.java]: Good idea. done.

**Comment by hustodemon** [comment]
Merged. Thanks for the review @ncounter !!!

**Pull request merged**